<html>
<body>
<link rel="stylesheet" type="text/css" media="all" href="css/common.css" />
<div id="mainContainer">
<?php require('header.html'); ?>
<div id="pageContents">
<?php require('navigation.html'); ?>

#Add your own FAQ questions and answers here

<p><b>Where did the name qTeller come from?</b></p>

<p>qTeller: The website that <i>tells</i> you all about the genes under you <i>QTL.</i> Hey, we had to call it something!</p>

<p><b>How are the expression values displayed by qTeller calculated?</b></p>

<p>The standard pipeline for qTeller uses GSNAP to align raw RNA sequencing reads against a reference genome or reference transcriptome. The expression levels of individual genes are quantified using <a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a> (for genomes) of <a href="http://bio.math.berkeley.edu/eXpress/overview.html">eXpress</a> for transcriptome data.</p>

<p><b>Why use GSNAP?</b></p>

<p>As the length of reads generated by Illumina continues to grow, more and more of these reads will span multiple exons. Many of the most popular short-read alignment software packages like Bowtie cannot handle aligning these spliced reads, which means the reads are simply thrown out as unalignable. This is a problem for two reasons: <ul><li>Throwing out a lot of reads means our measures of gene expression would be less accurate than they otherwise would be.</li><li>Genes with many small exons (instead of one or a couple of large exons) will disproportionately appear to lose expression as sequencing lengths get longer.</li></ul>
GSNAP is one of a handful of aligners that can carry out spliced alignments to a reference genome. A good comparison of some of the alternatives (for example TopHap, MapSplice, and RUM) is presented in this paper: "<a href="http://bioinformatics.oxfordjournals.org/content/early/2011/07/19/bioinformatics.btr427">Comparative Analysis of RNA-Seq Alignment Algorithms and the RNA-Seq Unified Mapper (RUM).</a>"</p>

<p><b>I have a bunch of RNA-seq data, but it isn't from maize, am I out of luck?</b></p>

<p>Not necessarily. The modular framework for both qTeller visualizations and the back-end RNA-seq analysis means it's possible to generate an instance of qTeller for practically any species with a reference genome. But it does take a certain critical mass of RNA-seq data before qTeller figures become informative. For an example of what is NOT enough data to really be informative, try generating bar plots or scatter plots using Arabidopsis thaliana gene names. <a href="http://qteller.com/qteller2/bar_chart.php?name=AT1G01010">Example.</a></p>

<p><b>What gave you the idea to create this tool?</b></p>
<p>The idea for this tool started when a friend asked me "Can you tell me where the genes under this QTL I'm interested in are expressed?" After the fourth or fifth time I was asked this question (or the related one "I'm really interested in gene X, can you tell me where it is expressed?") I figured I might as well build a web interface so I wouldn't need to look up the expression values for genes in specific regions manually each time. And if I was going to build I website anyway, I figured I would include all the other datasets I use myself when I'm trying to decide which genes I'm most interested in. In maize that means genes with syntenic orthologs in other grasses, and genes which have been previously characterized and studied -- the classical genes of maize genetics -- which are disproprotionately likely to <i>also</i> be genes with conserved syntenic orthologs in other grass species. </p>
<p>Then my boss asked me to put together another version of the website for a second species and after a lot of work messing around with re-customizing all of my scripts and web forms for that species I realized how powerful it'd be to make my tools modular so I could quickly analyze data and deploy and instance of qTeller for any set of data in any species. -James</p>

<p>If your question wasn't answered here, click "Contact" on the left hand side of the screen and get in touch. I'll try to get back to you as quickly as possible.</p>
</div>
</div>
</body>
</html>
